src/modules/CSADigitizer/CSADigitizerModule.cpp

```diff
@@ -0,0 +1,337 @@
+      gf_ = ikrum_/(2.0*1.5*vt_)  ;  
+      rf_ = 2./gf_;       //feedback resistor
+      tauF_=rf_*cf_;
+      tauR_=(cd_*co_)/(gm_ *cf_);
+      LOG(DEBUG) << "Parameters: rf " << Units::display(rf_, "V*s/C") << ", cf_ " << Units::display(cf_, "C/V")
+		 << ", cd_ " << Units::display(cd_, "C/V") << ", co_ " << Units::display(co_, "C/V")
+		 << ", gm_ " << Units::display(gm_, "C/s/V") << ", tauF_ " << Units::display(tauF_, "s")
+		 << ", tauR_ " << Units::display(tauR_, "s");      
+    }
+    
+    output_plots_ = config_.get<bool>("output_plots");
+    output_pulsegraphs_ = config_.get<bool>("output_pulsegraphs");
+
+
+    // impulse response of the CSA from Kleczek 2016 JINST11 C12001
+    // H(s) = Rf / ((1+ tau_f s) * (1 + tau_r s)), with
+    // tau_f = Rf Cf , rise time constant tau_r = (C_det * C_out) / ( gm_ * C_F )
+    // inverse Laplace transform R/((1+a*s)*(1+s*b)) is (wolfram alpha) (R (e^(-t/a) - e^(-t/b))) /(a - b) 
+    fImpulseResponse_ = new TF1("fImpulseResponse_", "[0] * ( exp(-x / [1]) - exp(-x/[2]) ) / ([1]-[2])", 0, tmax_);
+    fImpulseResponse_->SetParameters(rf_, tauF_, tauR_);
+
+
+}
+
+
+
+void CSADigitizerModule::init() {
+
+    if(config_.get<bool>("output_plots")) {
+        LOG(TRACE) << "Creating output plots";
+
+        // Plot axis are in kilo electrons - convert from framework units!
+        int maximum = static_cast<int>(Units::convert(config_.get<int>("output_plots_scale"), "ke"));
+        auto nbins = config_.get<int>("output_plots_bins");
+
+	//asv do these histos make sense? what about one with the noise?
+        // Create histograms if needed
+        h_pxq = new TH1D("pixelcharge", "raw pixel charge;pixel charge [ke];pixels", nbins, 0, maximum);
+        h_tot = new TH1D("tot", "time over threshold;time over threshold [ns];pixels", nbins, 0, maximum);
+        h_toa = new TH1D("toa", "time of arrival;time of arrival [ns];pixels", nbins, 0, maximum);
+	h_pxq_vs_tot = new TH2D("pxqvstot", "ToT vs raw pixel charge;pixel charge [ke];ToT [ns]", nbins, 0, maximum, nbins, 0, tmax_);
+
+    }
+
+    
+}
+
+
+
+void CSADigitizerModule::run(unsigned int event_num) {
+    // Loop through all pixels with charges
+    std::vector<PixelHit> hits;
+    for(auto& pixel_charge : pixel_message_->getData()) {
+        auto pixel = pixel_charge.getPixel();
+        auto pixel_index = pixel.getIndex();
+        auto inputcharge = static_cast<double>(pixel_charge.getCharge());
+
+        LOG(DEBUG) << "Received pixel " << pixel_index << ", charge " << Units::display(inputcharge, "e");
+
+	const auto& pulse = pixel_charge.getPulse(); // the pulse containing charges and times
+	auto pulse_vec = pulse.getPulse();           // the vector of the charges 
+        auto timestep = pulse.getBinning();
+	const int npx = static_cast<int>(ceil(tmax_/timestep));
+
+	if (first_event_){ // initialize impulse response function - assume all time bins are equal
```

Comment:
I guess you placed this here because in the `init` function we don't have the time binning yet? I'm still not a big fan and maybe we could find a different solution.

Reply:
Yes, that's why I did it. The only other option I could see was to make the timestep a config variable; and terminate the program with if it turns out during the first event that the binning in the input data is different than the one given in the config file. But I wasn't sure if that wouldn't be more frustrating for the user?

Reply:
Let's have a look at how this could best be solved once the other changes are in. But I agree that having to configure this is frustrating and I would prefer the current solution over that. But maybe we find a third one? :)